---
title: "Differential gene expression analysis"
author: "Dane Grosvenor"
date: "version `r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    toc: yes
    toc_float:
      collapsed: no
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Voom to DEG

### Load libraries

```{r load_libs, message=FALSE, warning=FALSE}

library(tidyverse)
library(edgeR)
library(limma)
library(kimma)

set.seed(101)

```

### Create output directories

```{r create_out_dirs, message=FALSE}

dir.create("results/lme_outputs/", showWarnings = FALSE, recursive = TRUE)

```

### Load Voom Object

```{r load_dat, message=FALSE}

# load RData w/ multiple objects
loadRData <- function(fileName) {
  load(fileName)
  get(ls()[ls() !="fileName"])
}

# https://stackoverflow.com/questions/5577221/how-can-i-load-an-object-into-a-variable-name-that-i-specify-from-an-r-data-file

# load voom object (see sh2b3_clean_voom.Rmd)
voom.obj <- loadRData("data_clean/voom2.RData")

# set Time and Genotype as a factor
voom.obj$targets$Genotype <- as.factor(voom.obj$targets$Genotype)
voom.obj$targets$Time <- as.factor(voom.obj$targets$Time)

voom.obj$targets <- voom.obj$targets %>%
  tibble::rownames_to_column(var = "ptID")

```

```{r include=FALSE}
# Load model results
load("results/lme_outputs/experiment1_lme.RData")
load("results/lme_outputs/experiment2_lme.RData")
```

### Create ensembl ID Gene symbol dictionary

```{r ensembl_symbol_dict, message=FALSE}

voom.obj$genes <- voom.obj$genes %>%
  dplyr::select(ensembl_gene_id, symbol) %>% # create ensembl:symbol dict
  tidyr::unnest(symbol) %>%
  # for ensembl IDs w/o associated gene symbols, write ensembl ID as gene symbol
  dplyr::mutate(symbol = as.list(case_when(
    ensembl_gene_id == "ENSMUSG00000062783" ~ "Csprs",
    ensembl_gene_id %in% c("ENSMUSG00000079190", "ENSMUSG00000079192", "ENSMUSG00000079222",
                           "ENSMUSG00000079808", "ENSMUSG00000095041", "ENSMUSG00000095250",
                           "ENSMUSG00000095672", "ENSMUSG00000095742") ~ ensembl_gene_id,
    TRUE ~ symbol)))

# single object, use saveRDS()
saveRDS(voom.obj$genes, 
     file = "data_clean/ensembl_ID_gene_symbol_dict.rds")

```

## Subset Experiments from Voom object

### Subset Experiment 1 from Voom object (Time as Factor)

```{r subset_exp1, message=FALSE}

# set experiment 1 as own object for downstream subset
voom.exp1 <- voom.obj

# filter experiment 1 (Expt == A) and corresponding time pts
voom.exp1$targets <- voom.exp1$targets %>%
  dplyr::filter(Expt == "A") %>%
  droplevels(.)

subset_sample_names_exp1 <- voom.exp1$targets$ptID

# add voom.exp1$weights colnames
colnames(voom.exp1$weights) <- colnames(voom.exp1$E)

# subset weight samples from target subset
voom.exp1$weights <- as.data.frame(voom.exp1$weights) %>%
  dplyr::select(all_of(subset_sample_names_exp1)) %>% 
  as.matrix()

# subset voom.exp1$E from target subset
voom.exp1$E <- as.data.frame(voom.exp1$E) %>%
  dplyr::select(all_of(subset_sample_names_exp1)) %>%
  as.matrix()

save(voom.exp1, file = "data_clean/voom_exp1.RData")

```

### Subset Experiment 2 from Voom object

```{r subset_exp2, message=FALSE}

# set experiment 2 as own object for downstream subset
voom.exp2 <- voom.obj

# filter experiment 2 (Expt == B) and remove smaller batch
voom.exp2$targets <- voom.exp2$targets %>%
  dplyr::filter(Expt == "B" & Batch == "2") %>%
  droplevels(.)

subset_sample_names_exp2 <- voom.exp2$targets$ptID

# add voom.exp2$weights colnames
colnames(voom.exp2$weights) <- colnames(voom.exp2$E)

# subset weight samples from target subset
voom.exp2$weights <- as.data.frame(voom.exp2$weights) %>%
  dplyr::select(all_of(subset_sample_names_exp2)) %>% 
  as.matrix()

# subset voom.exp2$E from target subset
voom.exp2$E <- as.data.frame(voom.exp2$E) %>%
  dplyr::select(all_of(subset_sample_names_exp2)) %>%
  as.matrix()

save(voom.exp2, file = "data_clean/voom_exp2.RData")

```

## Linear Mixed Effect Models

### Experiment 1 LME (Time as Factor)

```{r exp1_lme, eval=FALSE}

exp1.lme <- kimma::kmFit(dat = voom.exp1, patientID = "Animal",
                      model = "~ Time*Genotype + (1|Animal)",
                      contrast_var = "Time:Genotype",
                      run_lme = TRUE,
                      use_weights = TRUE,
                      run_contrast = TRUE,
                      metrics = TRUE)

save(exp1.lme, file = "results/lme_outputs/experiment1_lme.RData")

```

```{r}

kimma::summarise_kmFit(exp1.lme$lme,
                fdr.cutoff = c(0.05, 0.1, 0.25))

```

### Experiment 2 LME

```{r exp2_lme, eval=FALSE}

exp2.lme <- kimma::kmFit(dat = voom.exp2, patientID = "Animal",
                      model = "~ Time*Genotype + (1|Animal)",
                      contrast_var = "Time:Genotype",
                      run_lme = TRUE,
                      use_weights = TRUE,
                      run_contrast = TRUE,
                      metrics = TRUE)

save(exp2.lme, file = "results/lme_outputs/experiment2_lme.RData")

```

```{r}

kimma::summarise_kmFit(exp2.lme$lme,
                fdr.cutoff = c(0.05, 0.1, 0.25))

```

### Session Info

```{r sesh_info}

sessionInfo()

```



